glean-core/android/src/main/java/mozilla/telemetry/glean/acmigration/GleanACDataMigrator.kt

```diff
@@ -0,0 +1,163 @@
+            val values = IntArray(sequenceNumbers.size) { seqList[it].second }
+
+            return Pair(keys, values)
+        }
+    }
+
+    /**
+     * Get the metadata from glean-ac.
+     *
+     * @return an instance of [ACMetadata] with, at least, [ACMetadata.alreadyMigrated].
+     */
+    fun getACMetadata(): ACMetadata {
+        if (wasMigrated()) {
+            return ACMetadata(true, emptyMap())
+        }
+
+        return ACMetadata(
+            alreadyMigrated = false,
+            sequenceNumbers = getSequenceNumbers()
+        )
+    }
+
+    /**
+     * Load the sequence numbers that were stored in glean-ac [SharedPreferences].
+     *
+     * @return a `Map<String, Int>` in which entries are (`<ping_name>_seq`, nextSequenceNumber).
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getSequenceNumbers(): Map<String, Int> {
+        val seqPrefs = applicationContext.getSharedPreferences(
+            SEQUENCE_NUMBERS_FILENAME,
+            Context.MODE_PRIVATE
+        )
+
+        return try {
+            // The code in this block might throw. It probably means the file
+            // does not exist or the data is corrupted. We try our best to recover
+            // most of the data, though.
+            seqPrefs
+                .all
+                .entries
+                .filter {
+                    it.key.endsWith("_seq") &&
+                    it.value is Int &&
+                    it.value as Int >= 0
+                }
+                .map {
+                    val seq = it.value as Int
+                    it.key to seq
+                }.toMap()
+        } catch (e: NullPointerException) {
+            emptyMap()
+        }
+    }
+
+    /**
+     * Mark the current client as migrated.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun markAsMigrated() {
+        migrationPrefs?.edit()?.putBoolean("wasMigrated", true)?.apply()
+    }
+
+    /**
+     * Get if the current client was migrated.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun wasMigrated(): Boolean {
+        val defaultValue = false
+        return try {
+            migrationPrefs?.getBoolean("wasMigrated", defaultValue) ?: defaultValue
+        } catch (e: ClassCastException) {
+            // If another pref in this file exists with a non boolean value,
+            // something probably went wrong in the migration. Let's try again?
+            defaultValue
+        }
+    }
+
+    /**
+     * Check if data migration should be performed.
+     *
+     * @return true if glean-ac data was found and needs to be moved over,
+     *         false otherwise.
+     */
+    internal fun shouldMigrateData(): Boolean {
+        if (wasMigrated()) {
+            return false
+        }
+
+        // TODO: check if we have any previous a-c data (i.e. client_id). If so, migrate.
```

Comment:
Just as a check: will those todos land as follow-ups? Does it make sense to add their bug numbers?

Reply:
This was separated from this PR to make things easier to review. The client-id check will come with the 'generic engine' migration code. It will come right after this PR, otherwise nothing will work. No need to put a bug number here, I guess :)